File: src/state.rs
use borsh::{BorshDeserialize, BorshSerialize};
use num_derive::{FromPrimitive, ToPrimitive};
use solana_program::{
    account_info::AccountInfo,
    program_error::ProgramError,
    program_pack::{IsInitialized, Pack, Sealed},
    pubkey::Pubkey,
};
use std::{cell::RefCell, convert::TryInto, rc::Rc};

use crate::error::DexError;

#[derive(BorshDeserialize, BorshSerialize, Clone, Debug, PartialEq)]
#[allow(missing_docs)]
pub enum AccountTag {
    Initialized,
    DexState,
    UserAccount,
}

#[derive(BorshDeserialize, BorshSerialize, Clone, Copy, PartialEq, FromPrimitive, ToPrimitive)]
#[repr(u8)]
#[allow(missing_docs)]
pub enum Side {
    Bid,
    Ask,
}

impl Side {
    pub fn opposite(&self) -> Self {
        match self {
            Side::Bid => Side::Ask,
            Side::Ask => Side::Bid,
        }
    }
}

#[derive(BorshDeserialize, BorshSerialize, Clone, PartialEq)]
pub enum SelfTradeBehavior {
    DecrementTake,
    CancelProvide,
    AbortTransaction,
}

#[derive(BorshDeserialize, BorshSerialize)]
pub struct DexState {
    pub tag: AccountTag,
    pub signer_nonce: u8,
    pub base_mint: Pubkey,
    pub quote_mint: Pubkey,
    pub base_vault: Pubkey,
    pub quote_vault: Pubkey,
    pub orderbook: Pubkey,
}

impl DexState {
    pub(crate) fn check(self) -> Result<Self, ProgramError> {
        if self.tag != AccountTag::DexState {
            return Err(ProgramError::InvalidAccountData);
        };
        Ok(self)
    }
}

#[derive(BorshDeserialize, BorshSerialize)]
pub struct UserAccountHeader {
    pub tag: AccountTag,
    pub market: Pubkey,
    pub owner: Pubkey,
    pub base_token_free: u64,
    pub base_token_locked: u64,
    pub quote_token_free: u64,
    pub quote_token_locked: u64,
    pub number_of_orders: u32,
}

impl Sealed for UserAccountHeader {}

impl Pack for UserAccountHeader {
    const LEN: usize = 101;

    fn pack_into_slice(&self, mut dst: &mut [u8]) {
        self.serialize(&mut dst).unwrap()
    }

    fn unpack_from_slice(mut src: &[u8]) -> Result<Self, ProgramError> {
        UserAccountHeader::deserialize(&mut src).map_err(|_| ProgramError::InvalidAccountData)
    }
}

impl IsInitialized for UserAccountHeader {
    fn is_initialized(&self) -> bool {
        self.tag == AccountTag::UserAccount
    }
}

pub struct UserAccount<'a> {
    pub header: UserAccountHeader,
    data: Rc<RefCell<&'a mut [u8]>>,
}

impl<'a> UserAccount<'a> {
    pub fn parse(account: &AccountInfo<'a>) -> Result<Self, ProgramError> {
        Ok(Self {
            header: UserAccountHeader::unpack(&account.data.borrow())?,
            data: Rc::clone(&account.data),
        })
    }

    pub fn write(&self) {
        self.header.pack_into_slice(&mut self.data.borrow_mut());
    }

    pub fn read_order(&self, order_index: usize) -> Result<u128, DexError> {
        if order_index >= self.header.number_of_orders as usize {
            return Err(DexError::InvalidOrderIndex);
        }
        let offset = UserAccountHeader::LEN + order_index * 16;
        Ok(u128::from_le_bytes(
            self.data.borrow()[offset..offset + 16].try_into().unwrap(),
        ))
    }

    pub fn remove_order(&mut self, order_index: usize) -> Result<(), DexError> {
        if order_index >= self.header.number_of_orders as usize {
            return Err(DexError::InvalidOrderIndex);
        }
        if self.header.number_of_orders - order_index as u32 != 1 {
            let last_order = self.read_order((self.header.number_of_orders - 1) as usize)?;
            let offset = UserAccountHeader::LEN + order_index * 16;
            self.data.borrow_mut()[offset..offset + 16].copy_from_slice(&last_order.to_le_bytes());
        }
        self.header.number_of_orders -= 1;
        Ok(())
    }

    pub fn add_order(&mut self, order: u128) -> Result<(), DexError> {
        let offset = UserAccountHeader::LEN + (self.header.number_of_orders * 16) as usize;
        self.data
            .borrow_mut()
            .get_mut(offset..offset + 16)
            .map(|b| b.copy_from_slice(&order.to_le_bytes()))
            .ok_or(DexError::UserAccountFull)?;
        self.header.number_of_orders += 1;
        Ok(())
    }

    pub fn find_order_index(&self, order_id: u128) -> Result<usize, DexError> {
        let data: &[u8] = &self.data.borrow();
        Ok((UserAccountHeader::LEN..)
            .step_by(16)
            .take(self.header.number_of_orders as usize)
            .map(|offset| u128::from_le_bytes(data[offset..offset + 16].try_into().unwrap()))
            .enumerate()
            .find(|(_, b)| b == &order_id)
            .ok_or(DexError::OrderNotFound)?
            .0)
    }
}

pub trait Order {}

impl Order for u128 {}

File: js/src/state.ts
import { Connection, PublicKey } from "@solana/web3.js";
import BN from "bn.js";
import { deserialize, deserializeUnchecked, Schema } from "borsh";

export const CALLBACK_INFO_LEN = 33;

export enum AccountTag {
  Initialized = 0,
  MarketState = 1,
  UserAccount = 2,
}

export enum SelfTradeBehavior {
  DecrementTake = 0,
  CancelProvide = 1,
  AbortTransaction = 2,
}

export class MarketState {
  tag: AccountTag;
  baseMint: PublicKey;
  quoteMint: PublicKey;
  baseVault: PublicKey;
  quoteVault: PublicKey;
  orderbook: PublicKey;
  admin: PublicKey;
  creationTimestamp: BN;
  baseVolume: BN;
  quoteVolume: BN;
  accumulatedFees: BN;
  minBaseOrderSize: BN;
  signerNonce: number;
  feeTierThresholds: BN[];
  feeTierTakerBpsRates: BN[];
  feeTierMakerBpsRates: BN[];

  static schema: Schema = new Map([
    [
      MarketState,
      {
        kind: "struct",
        fields: [
          ["tag", "u64"],
          ["baseMint", [32]],
          ["quoteMint", [32]],
          ["baseVault", [32]],
          ["quoteVault", [32]],
          ["orderbook", [32]],
          ["admin", [32]],
          ["creationTimestamp", "u64"],
          ["baseVolume", "u64"],
          ["quoteVolume", "u64"],
          ["accumulatedFees", "u64"],
          ["minBaseOrderSize", "u64"],
          ["signerNonce", "u64"],
          ["feeTierThresholds", ["u64", 6]],
          ["feeTierTakerBpsRates", ["u64", 7]],
          ["feeTierMakerBpsRates", ["u64", 7]],
        ],
      },
    ],
  ]);

  constructor(obj: {
    tag: BN;
    signerNonce: BN;
    baseMint: Uint8Array;
    quoteMint: Uint8Array;
    baseVault: Uint8Array;
    quoteVault: Uint8Array;
    orderbook: Uint8Array;
    admin: Uint8Array;
    creationTimestamp: BN;
    baseVolume: BN;
    quoteVolume: BN;
    accumulatedFees: BN;
    minBaseOrderSize: BN;
    feeTierThresholds: BN[];
    feeTierTakerBpsRates: BN[];
    feeTierMakerBpsRates: BN[];
  }) {
    this.tag = obj.tag.toNumber() as AccountTag;
    this.signerNonce = obj.signerNonce.toNumber();
    this.baseMint = new PublicKey(obj.baseMint);
    this.quoteMint = new PublicKey(obj.quoteMint);
    this.baseVault = new PublicKey(obj.baseVault);
    this.quoteVault = new PublicKey(obj.quoteVault);
    this.orderbook = new PublicKey(obj.orderbook);
    this.admin = new PublicKey(obj.admin);
    this.creationTimestamp = obj.creationTimestamp;
    this.baseVolume = obj.baseVolume;
    this.quoteVolume = obj.quoteVolume;
    this.accumulatedFees = obj.accumulatedFees;
    this.minBaseOrderSize = obj.minBaseOrderSize;
    this.feeTierThresholds = obj.feeTierThresholds;
    this.feeTierTakerBpsRates = obj.feeTierTakerBpsRates;
    this.feeTierMakerBpsRates = obj.feeTierMakerBpsRates;
  }

  static async retrieve(connection: Connection, market: PublicKey) {
    const accountInfo = await connection.getAccountInfo(market);
    if (!accountInfo?.data) {
      throw new Error("Invalid account provided");
    }
    return deserialize(
      this.schema,
      MarketState,
      accountInfo.data
    ) as MarketState;
  }
}

export class UserAccount {
  tag: AccountTag;
  market: PublicKey;
  owner: PublicKey;
  baseTokenFree: BN;
  baseTokenLocked: BN;
  quoteTokenFree: BN;
  quoteTokenLocked: BN;
  accumulatedRebates: BN;
  accumulatedMakerQuoteVolume: BN;
  accumulatedMakerBaseVolume: BN;
  accumulatedTakerQuoteVolume: BN;
  accumulatedTakerBaseVolume: BN;
  orders: BN[];

  static schema: Schema = new Map([
    [
      UserAccount,
      {
        kind: "struct",
        fields: [
          ["tag", "u64"],
          ["market", [32]],
          ["owner", [32]],
          ["baseTokenFree", "u64"],
          ["baseTokenLocked", "u64"],
          ["quoteTokenFree", "u64"],
          ["quoteTokenLocked", "u64"],
          ["accumulatedRebates", "u64"],
          ["accumulatedMakerQuoteVolume", "u64"],
          ["accumulatedMakerBaseeVolume", "u64"],
          ["accumulatedTakerQuoteVolume", "u64"],
          ["accumulatedTakerBaseVolume", "u64"],
          ["_padding", "u32"],
          ["orders", ["u128"]],
        ],
      },
    ],
  ]);

  constructor(obj: {
    tag: BN;
    market: Uint8Array;
    owner: Uint8Array;
    baseTokenFree: BN;
    baseTokenLocked: BN;
    quoteTokenFree: BN;
    quoteTokenLocked: BN;
    orders: BN[];
    accumulatedRebates: BN;
    accumulatedMakerQuoteVolume: BN;
    accumulatedMakerBaseVolume: BN;
    accumulatedTakerQuoteVolume: BN;
    accumulatedTakerBaseVolume: BN;
  }) {
    this.tag = obj.tag.toNumber();
    this.market = new PublicKey(obj.market);
    this.owner = new PublicKey(obj.owner);
    this.baseTokenFree = obj.baseTokenFree;
    this.baseTokenLocked = obj.baseTokenLocked;
    this.quoteTokenFree = obj.quoteTokenFree;
    this.quoteTokenLocked = obj.quoteTokenLocked;
    this.orders = obj.orders;
    this.accumulatedRebates = obj.accumulatedRebates;
    this.accumulatedMakerQuoteVolume = obj.accumulatedMakerQuoteVolume;
    this.accumulatedMakerBaseVolume = obj.accumulatedMakerBaseVolume;
    this.accumulatedTakerQuoteVolume = obj.accumulatedTakerQuoteVolume;
    this.accumulatedTakerBaseVolume = obj.accumulatedTakerBaseVolume;
  }

  static async retrieve(connection: Connection, userAccount: PublicKey) {
    const accountInfo = await connection.getAccountInfo(userAccount);
    if (!accountInfo?.data) {
      throw new Error("Invalid account provided");
    }
    return deserializeUnchecked(
      this.schema,
      UserAccount,
      accountInfo.data
    ) as UserAccount;
  }
}
